libyul/optimiser/MemoryEscalator.cpp

```diff
@@ -0,0 +1,139 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <libyul/optimiser/MemoryEscalator.h>
+#include <libyul/optimiser/CallGraphGenerator.h>
+#include <libyul/optimiser/FunctionCallFinder.h>
+#include <libyul/optimiser/NameDispenser.h>
+#include <libyul/optimiser/StackToMemoryMover.h>
+#include <libyul/backends/evm/EVMDialect.h>
+#include <libyul/AsmData.h>
+#include <libyul/CompilabilityChecker.h>
+#include <libyul/Dialect.h>
+#include <libyul/Exceptions.h>
+#include <libyul/Object.h>
+#include <libyul/Utilities.h>
+#include <libsolutil/Algorithms.h>
+#include <libsolutil/CommonData.h>
+
+using namespace std;
+using namespace solidity;
+using namespace solidity::yul;
+
+MemoryOffsetAllocator::MemoryOffsetAllocator(
+	map<YulString, FunctionStackErrorInfo> const& _functionStackErrorInfo,
+	set<YulString> const& _functionsInCycle,
+	map<YulString, set<YulString>> const& _callGraph
+): m_functionStackErrorInfo(_functionStackErrorInfo), m_functionsInCycle(_functionsInCycle), m_callGraph(_callGraph)
+{
+}
+
+bool MemoryOffsetAllocator::run() { return run({}); }
+uint64_t MemoryOffsetAllocator::requiredSlots() const {	return m_nextAvailableSlot.at({}); }
+
+bool MemoryOffsetAllocator::run(YulString v)
+{
+	yulAssert(!m_nextAvailableSlot.count(v), "");
+	uint64_t& n = m_nextAvailableSlot[v];
```

Comment:
Is `m_nextAvailableSlot` the number of slots needed for function `v` (and its callees)?

Reply:
I think an explicit assignment at the end of the function would be clearer than dragging this single-letter reference across the whole function.

Reply:
Oh wow, this has the side-effect of the function not being called from a child on `v`, because it is set then, right? I would prefer to have this explicit instead by checking `m_functionsInCycle`...

Reply:
> Oh wow, this has the side-effect of the function not being called from a child on `v`, because it is set then, right? I would prefer to have this explicit instead by checking `m_functionsInCycle`...

Yeah, it does have that side-effect :-). I thought I had time until Monday to smooth this stuff over ;-).